File: src/ros2_bridge/builder/topic.rs
use std::sync::Arc;

use crate::errors::Result;
use crate::ros2_bridge::mapper::{Direction, TopicMapper};

use super::config::Ros2BridgeBuilder;
use super::{IntoTopicMapper, TopicQos};

/// After [`Ros2BridgeBuilder::from_ros`]: only [`FromRos::to_bus`].
pub struct FromRos {
    parent: Ros2BridgeBuilder,
    ros_topic: String,
    ros_qos: TopicQos,
}

/// After [`FromRos::to_bus`]: only [`FromRosToBus::mapper`].
pub struct FromRosToBus {
    parent: Ros2BridgeBuilder,
    ros_topic: String,
    ros_qos: TopicQos,
    bus_topic: String,
    bus_qos: TopicQos,
}

/// ROS2→bus route ready for [`.lazy()`](Ros2ToBusReady::lazy) / [`.add()`](Ros2ToBusReady::add).
pub struct Ros2ToBusReady {
    parent: Ros2BridgeBuilder,
    ros_topic: String,
    ros_qos: TopicQos,
    bus_topic: String,
    bus_qos: TopicQos,
    mapper: Arc<dyn TopicMapper>,
    lazy: bool,
}

/// After [`Ros2BridgeBuilder::from_bus`]: only [`FromBus::to_ros`].
pub struct FromBus {
    parent: Ros2BridgeBuilder,
    bus_topic: String,
    bus_qos: TopicQos,
}

/// After [`FromBus::to_ros`]: only [`FromBusToRos::mapper`].
pub struct FromBusToRos {
    parent: Ros2BridgeBuilder,
    ros_topic: String,
    ros_qos: TopicQos,
    bus_topic: String,
    bus_qos: TopicQos,
}

/// Bus→ROS2 route ready for [`.add()`](BusToRos2Ready::add).
pub struct BusToRos2Ready {
    parent: Ros2BridgeBuilder,
    ros_topic: String,
    ros_qos: TopicQos,
    bus_topic: String,
    bus_qos: TopicQos,
    mapper: Arc<dyn TopicMapper>,
}

impl Ros2BridgeBuilder {
    pub fn from_ros(self, ros_topic: impl Into<String>, ros_qos: TopicQos) -> FromRos {
        FromRos {
            parent: self,
            ros_topic: ros_topic.into(),
            ros_qos,
        }
    }

    pub fn from_bus(self, bus_topic: impl Into<String>, bus_qos: TopicQos) -> FromBus {
        FromBus {
            parent: self,
            bus_topic: bus_topic.into(),
            bus_qos,
        }
    }
}

impl FromRos {
    pub fn to_bus(self, bus_topic: impl Into<String>, bus_qos: TopicQos) -> FromRosToBus {
        FromRosToBus {
            parent: self.parent,
            ros_topic: self.ros_topic,
            ros_qos: self.ros_qos,
            bus_topic: bus_topic.into(),
            bus_qos,
        }
    }
}

impl FromRosToBus {
    pub fn mapper(self, mapper: impl IntoTopicMapper) -> Ros2ToBusReady {
        Ros2ToBusReady {
            parent: self.parent,
            ros_topic: self.ros_topic,
            ros_qos: self.ros_qos,
            bus_topic: self.bus_topic,
            bus_qos: self.bus_qos,
            mapper: mapper.into_topic_mapper(),
            lazy: false,
        }
    }
}

impl Ros2ToBusReady {
    /// Opt-in lazy ROS 2 subscription. Default is eager at `build()`.
    pub fn lazy(mut self) -> Self {
        self.lazy = true;
        self
    }

    pub fn add(self) -> Result<Ros2BridgeBuilder> {
        self.parent.push_route(
            self.ros_topic,
            self.bus_topic,
            self.mapper,
            Direction::Ros2ToBus,
            self.lazy,
            self.ros_qos,
            self.bus_qos,
        )
    }
}

impl FromBus {
    pub fn to_ros(self, ros_topic: impl Into<String>, ros_qos: TopicQos) -> FromBusToRos {
        FromBusToRos {
            parent: self.parent,
            ros_topic: ros_topic.into(),
            ros_qos,
            bus_topic: self.bus_topic,
            bus_qos: self.bus_qos,
        }
    }
}

impl FromBusToRos {
    pub fn mapper(self, mapper: impl IntoTopicMapper) -> BusToRos2Ready {
        BusToRos2Ready {
            parent: self.parent,
            ros_topic: self.ros_topic,
            ros_qos: self.ros_qos,
            bus_topic: self.bus_topic,
            bus_qos: self.bus_qos,
            mapper: mapper.into_topic_mapper(),
        }
    }
}

impl BusToRos2Ready {
    pub fn add(self) -> Result<Ros2BridgeBuilder> {
        self.parent.push_route(
            self.ros_topic,
            self.bus_topic,
            self.mapper,
            Direction::BusToRos2,
            false,
            self.ros_qos,
            self.bus_qos,
        )
    }
}

File: bindings/python/robot_bus/ros2_bridge/builder/topic.py
"""Topic fluent stages for Ros2BridgeBuilder."""

from __future__ import annotations

from typing import Any

from .config import (
    Direction,
    Ros2BridgeBuilder,
    TopicQos,
    _require_bus_best_effort,
    _require_topic_qos,
    _topic_supports_lazy,
)

class FromRos:
    def __init__(self, parent: Ros2BridgeBuilder, ros_topic: str, ros_qos: TopicQos) -> None:
        self._parent = parent
        self._ros_topic = ros_topic
        self._ros_qos = ros_qos

    def to_bus(self, bus_topic: str, bus_qos: TopicQos) -> "FromRosToBus":
        qos = _require_topic_qos(bus_qos, "to_bus")
        _require_bus_best_effort(qos)
        return FromRosToBus(self._parent, self._ros_topic, self._ros_qos, bus_topic, qos)


class FromRosToBus:
    def __init__(
        self,
        parent: Ros2BridgeBuilder,
        ros_topic: str,
        ros_qos: TopicQos,
        bus_topic: str,
        bus_qos: TopicQos,
    ) -> None:
        self._parent = parent
        self._ros_topic = ros_topic
        self._ros_qos = ros_qos
        self._bus_topic = bus_topic
        self._bus_qos = bus_qos

    def mapper(self, mapper: Any) -> "Ros2ToBusReady":
        if mapper is None:
            raise ValueError("ros2 bridge route: mapper must not be None")
        return Ros2ToBusReady(
            self._parent,
            self._ros_topic,
            self._ros_qos,
            self._bus_topic,
            self._bus_qos,
            mapper,
        )


class Ros2ToBusReady:
    def __init__(
        self,
        parent: Ros2BridgeBuilder,
        ros_topic: str,
        ros_qos: TopicQos,
        bus_topic: str,
        bus_qos: TopicQos,
        mapper: Any,
    ) -> None:
        self._parent = parent
        self._ros_topic = ros_topic
        self._ros_qos = ros_qos
        self._bus_topic = bus_topic
        self._bus_qos = bus_qos
        self._mapper = mapper
        self._lazy = False

    def lazy(self) -> "Ros2ToBusReady":
        self._lazy = True
        return self

    def add(self) -> Ros2BridgeBuilder:
        if self._lazy and not _topic_supports_lazy(self._mapper):
            raise ValueError(
                "ros2 bridge route: .lazy() is not supported for this custom TopicMapper "
                "(attach-only); implement ros_msg_type/ros_to_bus"
            )
        self._parent._routes.append(
            {
                "ros_topic": self._ros_topic,
                "bus_topic": self._bus_topic,
                "mapper": self._mapper,
                "direction": Direction.Ros2ToBus,
                "lazy": self._lazy,
                "ros_qos": self._ros_qos,
                "bus_qos": self._bus_qos,
            }
        )
        return self._parent


class FromBus:
    def __init__(self, parent: Ros2BridgeBuilder, bus_topic: str, bus_qos: TopicQos) -> None:
        self._parent = parent
        self._bus_topic = bus_topic
        self._bus_qos = bus_qos

    def to_ros(self, ros_topic: str, ros_qos: TopicQos) -> "FromBusToRos":
        return FromBusToRos(
            self._parent,
            ros_topic,
            _require_topic_qos(ros_qos, "to_ros"),
            self._bus_topic,
            self._bus_qos,
        )


class FromBusToRos:
    def __init__(
        self,
        parent: Ros2BridgeBuilder,
        ros_topic: str,
        ros_qos: TopicQos,
        bus_topic: str,
        bus_qos: TopicQos,
    ) -> None:
        self._parent = parent
        self._ros_topic = ros_topic
        self._ros_qos = ros_qos
        self._bus_topic = bus_topic
        self._bus_qos = bus_qos

    def mapper(self, mapper: Any) -> "BusToRos2Ready":
        if mapper is None:
            raise ValueError("ros2 bridge route: mapper must not be None")
        return BusToRos2Ready(
            self._parent,
            self._ros_topic,
            self._ros_qos,
            self._bus_topic,
            self._bus_qos,
            mapper,
        )


class BusToRos2Ready:
    def __init__(
        self,
        parent: Ros2BridgeBuilder,
        ros_topic: str,
        ros_qos: TopicQos,
        bus_topic: str,
        bus_qos: TopicQos,
        mapper: Any,
    ) -> None:
        self._parent = parent
        self._ros_topic = ros_topic
        self._ros_qos = ros_qos
        self._bus_topic = bus_topic
        self._bus_qos = bus_qos
        self._mapper = mapper

    def add(self) -> Ros2BridgeBuilder:
        self._parent._routes.append(
            {
                "ros_topic": self._ros_topic,
                "bus_topic": self._bus_topic,
                "mapper": self._mapper,
                "direction": Direction.BusToRos2,
                "lazy": False,
                "ros_qos": self._ros_qos,
                "bus_qos": self._bus_qos,
            }
        )
        return self._parent

